[paradoxalarm] Paradox Alarm Binding initial contribution

theater opened the pull request:
[Paradox alarm system binding] [WIP] Initial contribution with initial scope

Description:
This is a new addon for binding directly to Paradox alarm systems via their IP150 (ethernet) interface using low-level byte-code packets.
Currently supported are only Evo type of panels as I own Evo192, but code is implemented with the idea of extending for other types as well if there are contributors who can do the low-level byte-code communication extension.

Initial discussion has been done here:
https://community.openhab.org/t/paradox-evo-binding/58026

Jar can be found here:
https://github.com/theater/binaries/tree/master/ParadoxAlarm


Question for reviewers:
I'm not sure if it's needed to add any disclaimer that the binding has nothing to do with Paradox company trademarks, etc. The implementation is done by reverse-engineering the byte code by other contributors who made the initial implementation in python and C#.

Could you please advise if the header on each class is sufficient or I need to add something more into the README.md?

Best regards,
Konstantin

Confectrician commented:
Maybe a bit early for a readme comment, but:

I have seen lots of html tags in your readme file, which are not needed or (which would be even worse) are doing strange things to the addon documentation site, when the openhab website is built and deployed.

We grab everything during the website built process and transform the markdown files to html pages on the fly.
So no need for html in most of all cases. (One exception would be large nested tables.)

theater commented:
Hi,
I agree and I would get rid of them but the issue I have is that the autoformatter on save in the eclipse deletes the double spaces at the end. Do you use separate editor for the MD files or somehow change this behavior only for .md files?

Cheers,
K.

Confectrician commented:
Sorry i can't help with the eclipse "problem". I am no eclipse expert. Maybe one of the maintainers has an idea for this problem.

theater commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/thing/zone.xml:
```diff
@@ -0,0 +1,54 @@
+
+		<channels>
+			<channel id="label" typeId="zoneLabel" />
+			<channel id="isOpened" typeId="isOpenedState" />
```
OK. Agree with proposed changes for is, are, has prefixes. Should I also hange the typeIds?

Hilbrand replied:
Yes preferable

theater replied:
Fixed with #4762

theater commented on addons/binding/org.openhab.binding.paradoxalarm/README.md:
```diff
@@ -0,0 +1,133 @@
+
+### IP150 parameters
+
+| Parameter     | Description                            |
```
How about when the description is too long, or we talk only for the header bars?

theater commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/thing/panel.xml:
```diff
@@ -0,0 +1,63 @@
+
+		<channels>
+			<channel id="state" typeId="panelState" />
+			<channel id="serialNumber" typeId="serialNumber" />
```
Could you elaborate? Do you mean properties that are only logged in the log during initial connection or there is other way to present them to the framework, aside from channels? I.e. if someone would like to put them in String items for example.

Hilbrand replied:
You can add them as `properties` on a thing. They are stored with the thing and for example shown in PaperUI when you show the thing.These can be set when you create a discovery result or edit them in the thing handler. If you need the value on an item you should create a channel indeed. But do you have a real use case or is it something that a user might do?

theater replied:
Agreed. Will make them properties.

theater commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/thing/zone.xml:
```diff
@@ -0,0 +1,54 @@
+		<state readOnly="true" pattern="%s" />
+	</channel-type>
+	<channel-type id="isTamperedState">
+		<item-type>boolean</item-type>
```
Hm... I'm not sure about this. Boolean represents best these states because it says true - is tampered, or false - not tampered. 
While Contact will permit easier graphical presentation of these states (green/red for ex.) I'm not sure if it's natural to bind them to OPEN/CLOSED state. Maybe we should use String here. What do you think?

theater commented on addons/binding/org.openhab.binding.paradoxalarm/.gitignore:
```diff
@@ -0,0 +1,2 @@
+target/*
```
Fixed #4762

theater commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/i18n/paradoxalarm_xx_XX.properties:
```diff
@@ -0,0 +1,13 @@
+# FIXME: please substitute the xx_XX with a proper locale, ie. de_DE
```
Fixed #4762

theater commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/thing/ip150connector.xml:
```diff
@@ -0,0 +1,52 @@
+
+		<config-description>
+			<parameter name="refresh" type="integer" min="1" max="600">
+				<label>Refresh interval</label>
```
Fixed #4762

theater commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/thing/panel.xml:
```diff
@@ -0,0 +1,63 @@
+		</channels>
+	</thing-type>
+
+    <channel-type id="panelState">
```
Fixed #4762

theater commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/thing/panel.xml:
```diff
@@ -0,0 +1,63 @@
+
+    <channel-type id="panelState">
+        <item-type>String</item-type>
+        <label>Panel state</label>
```
Fixed #4762

theater commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/thing/panel.xml:
```diff
@@ -0,0 +1,63 @@
+    </channel-type>
+	<channel-type id="serialNumber">
+		<item-type>String</item-type>
+		<label>Serial number</label>
```
Fixed #4762

theater commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/thing/panel.xml:
```diff
@@ -0,0 +1,63 @@
+	</channel-type>
+	<channel-type id="panelType">
+		<item-type>String</item-type>
+		<label>Panel type</label>
```
Fixed #4762

theater commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/thing/panel.xml:
```diff
@@ -0,0 +1,63 @@
+	</channel-type>
+	<channel-type id="applicationVersion">
+		<item-type>String</item-type>
+		<label>Application version</label>
```
Fixed with #4762

theater commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/thing/zone.xml:
```diff
@@ -0,0 +1,54 @@
+		<state readOnly="true" pattern="%s" />
+	</channel-type>
+	<channel-type id="isOpenedState">
+		<item-type>contact</item-type>
```
Fixed with #4762

theater commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/thing/zone.xml:
```diff
@@ -0,0 +1,54 @@
+			<channel id="label" typeId="zoneLabel" />
+			<channel id="isOpened" typeId="isOpenedState" />
+			<channel id="isTampered" typeId="isTamperedState" />
+			<channel id="hasLowBattery" typeId="hasLowBatteryState" />
```
Fixed with #4762

theater commented on addons/binding/org.openhab.binding.paradoxalarm/README.md:
```diff
@@ -0,0 +1,133 @@
+| Parameter     | Description                            |
+|---------------|---------------------------------------|
+| refresh       | Value is in seconds. Defines the refresh interval when the binding polls from paradox system.|
+| ip150Password |The password to your IP150 (not your panel PIN).|
```
Fixed with #4762

theater commented on addons/binding/pom.xml:
```diff
@@ -136,6 +136,7 @@
     <module>org.openhab.binding.openweathermap</module>
     <module>org.openhab.binding.openuv</module>
     <module>org.openhab.binding.orvibo</module>
+	<module>org.openhab.binding.paradoxalarm</module>
```
Hope it's fine now but on my IDE looked with OK identation. Deleted and did a single tab. Fixed with #4762

theater commented on features/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -553,6 +553,12 @@
+	
+	<feature name="openhab-binding-paradoxalarm" description="Paradox Alarm System Binding" version="${project.version}">
+        <feature>openhab-runtime-base</feature>
+        <feature>openhab-transport-serial</feature>
```
Fixed with #4762

marcelrv commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/thing/panel.xml:
```diff
@@ -0,0 +1,63 @@
+			<channel id="state" typeId="panelState" />
+			<channel id="serialNumber" typeId="serialNumber" />
+			<channel id="panelType" typeId="panelType" />
+			<channel id="hardwareVersion" typeId="hardwareVersion" />
```
as @Hilbrand indicated, these should be defined as properties
https://www.eclipse.org/smarthome/documentation/development/bindings/thing-definition.html#properties

theater replied:
Agree and also duplicate requirement. Will track that in Hilbrand's comment.

Hilbrand commented:
I think your last commit mangled the feature.xml (and still contains the incorrect indentation).

marcelrv commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/EvoCommunicator.java:
```diff
@@ -0,0 +1,336 @@
+    }
+
+    public String readPartitionLabel(int partitionNo) throws Exception {
+        logger.debug("Reading partition label: " + partitionNo);
```
Suggest instead
logger.debug("Reading partition label:  {}" , partitionNo);

theater replied:
Yes. Of course. Will resolve this in some of next commits.

marcelrv commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/Main.java:
```diff
@@ -0,0 +1,135 @@
+
+    private static Logger logger = LoggerFactory.getLogger(Main.class);
+
+    private static final String IP_ADDRESS = "192.168.254.231";
```
Is this a default of the device, or something in your own network?

theater replied:
Main class is for testing only but it's worth to extract IP address as one of the vararg parameters of the main class. Will be fixed. Thanks.

Hilbrand replied:
This comment reminded me I needed to do some more reviewing. You should remove this Main class in total. If you want to have some test, it's better to create a real unit test. But you need to create a separate project for it, so that' a little bit more work.

theater replied:
Unless it's strictly forbidden to have main class I would prefer to keep it because it eases the debugging of core low-level logic and tranlation into the model without the need to power up openhab and debug using remote aplication with the full stack.

marcelrv commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxPartitionHandler.java:
```diff
@@ -0,0 +1,58 @@
+ */
+public class ParadoxPartitionHandler extends EntityBaseHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(ParadoxPartitionHandler.class);
```
make this static

Hilbrand replied:
loggers are not made static in openHAB

theater replied:
All loggers will be made non-static as per Hilbrand's proposal even though I'm not sure it's good when creating multiple handlers for different zones which pretty much look the same.

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/parsers/IParadoxParser.java:
```diff
@@ -0,0 +1,35 @@
+ * @author Konstantin_Polihronov - Initial contribution
+ */
+public interface IParadoxParser {
+    public PartitionState calculatePartitionState(byte[] partitionFlags);
```
Will review all interfaces for such leftovers from extracting from the class. Thanks.

theater commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/thing/panel.xml:
```diff
@@ -0,0 +1,63 @@
+	</channel-type>
+	<channel-type id="hardwareVersion">
+		<item-type>String</item-type>
+		<label>Hardware version</label>
```
Fixed

theater commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/thing/panel.xml:
```diff
@@ -0,0 +1,63 @@
+	</channel-type>
+	<channel-type id="bootloaderVersion">
+		<item-type>String</item-type>
+		<label>Boot loader version</label>
```
Fixed

theater commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/thing/zone.xml:
```diff
@@ -0,0 +1,54 @@
+		<channels>
+			<channel id="label" typeId="zoneLabel" />
+			<channel id="isOpened" typeId="isOpenedState" />
+			<channel id="isTampered" typeId="isTamperedState" />
```
fixed

theater commented on addons/binding/org.openhab.binding.paradoxalarm/README.md:
```diff
@@ -0,0 +1,133 @@
+With the power of OpenHab this binding can be used for complex decision rules combining motion/magnetic sensor or whole partitions states with different scenarios.
+
+
+Examples: All partitions are armed, therefore there is no one at home. Window is opened for more than 10 minutes and temperature outside is bellow XXX degrees, send mail/any other supported notification to particular people.
```
Agree. Fixed with last commit.

theater commented on addons/binding/org.openhab.binding.paradoxalarm/build.properties:
```diff
@@ -0,0 +1,8 @@
+               OSGI-INF/,\
+               ESH-INF/,\
+               NOTICE
+jars.compile.order = .
```
Fixed.

theater commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/thing/ip150connector.xml:
```diff
@@ -0,0 +1,52 @@
+				<description>PanelType</description>
+			</parameter>
+			<parameter name="ip150Password" type="text" required="true">
+				<label>IP150 Password</label>
```
Fixed for the password and also probably is good for the PIN which is kinda 4 digit password. Is that OK?

Hilbrand replied:
yes probably a good idea.

theater commented on addons/binding/org.openhab.binding.paradoxalarm/.project:
```diff
@@ -0,0 +1,39 @@
+			</arguments>
+		</buildCommand>
+		<buildCommand>
+			<name>org.eclipse.m2e.core.maven2Builder</name>
```
done

theater commented on addons/binding/org.openhab.binding.paradoxalarm/.project:
```diff
@@ -0,0 +1,39 @@
+		</buildCommand>
+	</buildSpec>
+	<natures>
+		<nature>org.eclipse.m2e.core.maven2Nature</nature>
```
done

theater commented on addons/binding/org.openhab.binding.paradoxalarm/.settings/org.eclipse.core.resources.prefs:
```diff
@@ -0,0 +1,2 @@
+eclipse.preferences.version=1
```
done

theater commented on addons/binding/org.openhab.binding.paradoxalarm/.settings/org.eclipse.m2e.core.prefs:
```diff
@@ -0,0 +1,4 @@
+activeProfiles=
```
done

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/HandlersUtil.java:
```diff
@@ -0,0 +1,37 @@
+ */
+public class HandlersUtil {
+
+    private final static Logger logger = LoggerFactory.getLogger(HandlersUtil.class);
```
Is that fine if particular handlers tend to be created multiple times (like zones which may have 100+ handlers), to use static logger? 
Is it mandatory to use non-static loggers inside the model code whereever there is no static call to the logger?

theater replied:
This is not possible in HandlersUtil as logger is called statically.

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/ParadoxCommunicatorFactory.java:
```diff
@@ -0,0 +1,58 @@
+ */
+public class ParadoxCommunicatorFactory {
+
+    protected static Logger logger = LoggerFactory.getLogger(ParadoxCommunicatorFactory.class);
```
OK

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/OpenClosedType.java:
```diff
@@ -0,0 +1,41 @@
+ *
+ * @author Konstantin_Polihronov - Initial contribution
+ */
+public enum OpenClosedType implements State {
```
Actually it works. I use it in ParadoxZoneHandler but I agree that if it's used only there it makes sense to delete it and implement the boolean to open/closed state logic in a method. Not sure why I went into the enum implemenst state direction when I coded it....

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/messages/HeaderMessageType.java:
```diff
@@ -0,0 +1,40 @@
+ *
+ * @author Konstantin_Polihronov - Initial contribution
+ */
+
```
OK

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/HandlersUtil.java:
```diff
@@ -0,0 +1,37 @@
+        if (schedule != null) {
+            boolean cancelingResult = schedule.cancel(true);
+            String cancelingSuccessful = cancelingResult ? "successful" : "failed";
+            logger.debug("Canceling schedule of {} is {}", schedule.toString(), cancelingSuccessful);
```
OK

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/EntityBaseHandler.java:
```diff
@@ -0,0 +1,95 @@
+    protected abstract void updateEntity();
+
+    protected int calculateEntityIndex() {
+        int index = config.getId() - 1;
```
Done. Thanks for the idea !

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/EvoCommunicator.java:
```diff
@@ -0,0 +1,336 @@
+        initializeMemoryMap();
+    }
+
+    /*
```
done

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/EvoCommunicator.java:
```diff
@@ -0,0 +1,336 @@
+                result.add(readPartitionLabel(i));
+            }
+        } catch (Exception e) {
+            logger.debug("Unable to retrieve partition labels.\nException: " + e.getMessage());
```
fixed

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java:
```diff
@@ -0,0 +1,254 @@
+    private static final int INITIAL_SCHEDULE_DELAY = 15;
+    private static final int FAILED_READ_ATTEMPTS_TRESHOLD = 10;
+
+    private final static Logger logger = LoggerFactory.getLogger(ParadoxIP150BridgeHandler.class);
```
done

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java:
```diff
@@ -0,0 +1,254 @@
+    public ParadoxIP150BridgeHandler(Bridge bridge) throws Exception {
+        super(bridge);
+        logger.info("Starting creation of communicator handler");
+        config = getConfigAs(ParadoxIP150BridgeConfiguration.class);
```
OK. Done.

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java:
```diff
@@ -0,0 +1,254 @@
+
+    private void scheduleRefresh() {
+        logger.debug("Scheduling cache update. Refresh interval: " + config.getRefresh() + "s.");
+        refreshCacheUpdateSchedule = scheduler.scheduleWithFixedDelay(() -> {
```
Done. Thanks !

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxPanelHandler.java:
```diff
@@ -0,0 +1,104 @@
+                        new StringType(panelInformation.getBootLoaderVersion().toString()));
+            }
+        } catch (ParadoxBindingException e) {
+            logger.error("Unable to retrieve ParadoxPanel instance. {}", e);
```
Fixed everywhere but the main class.

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/util/ParadoxUtil.java:
```diff
@@ -0,0 +1,143 @@
+        if (array == null) {
+            return new String();
+        }
+        String result = DatatypeConverter.printHexBinary(array);
```
agree

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java:
```diff
@@ -0,0 +1,254 @@
+    }
+
+    private IParadoxCommunicator initializeCommunicator() throws Exception {
+        synchronized (ParadoxIP150BridgeHandler.class) {
```
Made the method synchronized as this is the expected behavior

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java:
```diff
@@ -0,0 +1,254 @@
+ *
+ * @author Konstantin_Polihronov - Initial contribution
+ */
+public class ParadoxIP150BridgeHandler extends BaseBridgeHandler {
```
OK I just did that but the default settings in IDE show my fields which are initialized in initialize method as errors with "can be null". Is there a way to handle this without lowering the severity for the issue?

Hilbrand replied:
Ok I meant use `@NonNullByDefault({})` on the fields. In the class it should be `@NonNullByDefault`.
So use `private @NonNullByDefault({}) ScheduledFuture<?> refreshCacheUpdateSchedule;`

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/model/PanelType.java:
```diff
@@ -0,0 +1,59 @@
+    public static PanelType from(String panelTypeStr) {
+        try {
+            return PanelType.valueOf(panelTypeStr);
+        } catch (Exception e) {
```
Agree. Event though in this case it doesn't really matter. To be fixed in next commit.

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/GenericCommunicator.java:
```diff
@@ -0,0 +1,329 @@
+    }
+
+    private void reinitializeSocket() throws UnknownHostException, IOException {
+        socket = new Socket(ipAddress, tcpPort);
```
Thanks for pointing that out ! Probably would be enough to check socket == null || socket.isClosed(). The others are created using the new socket. Is this what you mean?

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/util/ParadoxUtil.java:
```diff
@@ -0,0 +1,142 @@
+ */
+public class ParadoxUtil {
+
+    private static Logger logger = LoggerFactory.getLogger(ParadoxUtil.class);
```
```suggestion
    private final Logger logger = LoggerFactory.getLogger(ParadoxUtil.class);
```

theater replied:
I'll make it "static final" instead because it's called in a static way

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/util/ParadoxUtil.java:
```diff
@@ -0,0 +1,142 @@
+
+    public static String byteArrayAsString(byte[] array) {
+        if (array == null) {
+            return new String();
```
```suggestion
            return "";
```

Hilbrand commented on features/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -1,1055 +1,1070 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Something went wrong here. Better replace it with the original file and add the changes again.

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/EvoCommunicator.java:
```diff
@@ -0,0 +1,326 @@
+            for (int i = 1; i <= 8; i++) {
+                result.add(readPartitionLabel(i));
+            }
+        } catch (Exception e) {
```
Is it necessary to catch `Exception` here? Only catch specific exceptions.

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/EvoCommunicator.java:
```diff
@@ -0,0 +1,326 @@
+    public String readPartitionLabel(int partitionNo) throws Exception {
+        logger.debug("Reading partition label: " + partitionNo);
+        if (partitionNo < 1 || partitionNo > 8) {
+            throw new Exception("Invalid partition number. Valid values are 1-8.");
```
Don't throw this generic Exception. Either create a new exception or see if an existing exception fit.

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/EvoCommunicator.java:
```diff
@@ -0,0 +1,326 @@
+    }
+
+    public String readZoneLabel(int zoneNumber) throws Exception {
+        logger.debug("Reading zone label: " + zoneNumber);
```
```suggestion
        logger.debug("Reading zone label: {}", zoneNumber);
```
Note: Also check other log usages to not concatenate strings. Concatenating is always done even when the specific log level is not enabled, while using the variables as argument it's only evaluated when the specific log level is enabled.

theater replied:
I understand this. Looked everywhere. Hopefully with next commit it will be OK.

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/messages/RamRequestPayload.java:
```diff
@@ -0,0 +1,35 @@
+    @Override
+    protected byte calculateControlByte() {
+        byte byteValue = (byte) 0;
+        return ParadoxUtil.setBit(byteValue, 7, 1);
```
This looks like it's a constant. So you might just create a static field of it so it's not calculated every time. For example:
`private static final byte CONTROL_BYTE = ParadoxUtil.setBit((byte) 0, 7, 1);`
See if this pattern can be applied in other places in your code.

theater replied:
The method will return static pre-calculated constatn in this class. Thanks !

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/exceptions/ParadoxBindingException.java:
```diff
@@ -0,0 +1,50 @@
+    public ParadoxBindingException(String message, Throwable cause, boolean enableSuppression,
+            boolean writableStackTrace) {
+        super(message, cause, enableSuppression, writableStackTrace);
+        // TODO Auto-generated constructor stub
```
Remove generated comments. See also other palces.

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/exceptions/ParadoxBindingException.java:
```diff
@@ -0,0 +1,50 @@
+ */
+public class ParadoxBindingException extends Exception {
+
+    private static final long serialVersionUID = 1L;
```
if set use the generation of an actual value (in eclipse) instead of the value 1.

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/EntityBaseHandler.java:
```diff
@@ -0,0 +1,91 @@
+            }
+            logger.debug("Finished initializing!");
+        } catch (ParadoxBindingException e) {
+            logger.warn("Unable to retrieve/create Paradox panel instance. Exception: {}", e);
```
Maybe also set the status here.

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/model/PartitionState.java:
```diff
@@ -0,0 +1,327 @@
+            append(sb, "All zones closed");
+        }
+
+        // if (sb.substring(sb.length() - 2, sb.length() - 1).equals(";")) {
```
Is this code still relevant or can it be removed?

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/messages/ParadoxIPPacket.java:
```diff
@@ -0,0 +1,196 @@
+            this.payloadLength = (short) payload.length;
+        }
+
+        // TODO: Figure out how to fill up to 16, 32, 48, etc sizes with 0xEE
```
So far  I haven't seen any issues with not doing it but according to the guy who reverse-engineered the stuff this is how the packets are created from the original tools by Paradox.

theater commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java:
```diff
@@ -0,0 +1,255 @@
+        try {
+            logger.info("Starting creation of communicator handler");
+            config = getConfigAs(ParadoxIP150BridgeConfiguration.class);
+            communicator = initializeCommunicator();
```
Does that mean that I can keep the bridge OFFLINE until the scheduled task is being executed successfully? Is that OK? I guess you advice me to just read the config, set the fields and continue offline until the execution of connect to the system happens...

Hilbrand replied:
Yes that is how it done in most cases. Set the bridge offline or unknown and when the actual device information is received set the state based on that information.

theater replied:
Just want to let you all know that I'm currently refactoring the whole lifecycle and execution of handlers to use async methods. Need more decoupling though so it will take time before the actual commit...

theater replied:
Everything should have been fixed now. Looks OK to me. Please review.

cweitkamp commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/util/ParadoxUtil.java:
```diff
@@ -0,0 +1,142 @@
+        if (array == null) {
+            return "";
+        }
+        return DatatypeConverter.printHexBinary(array);
```
Please replace the usage of `DatatypeConverter.printHexBinary()` with the `HexUtils.bytesToHex()` method provided by the framework.

theater replied:
Nice that we have it in the framework. I refactored all calls recursively and removed completely the method including subsequent calls.

theater commented:
I reverted to the commit before my initial commit: ceb6f8d2ecc3597cd57a1ea1c18189ea46da3883 Not sure why I have conflicts. Can someone help with that?

theater commented:
OK. I've just added again my binding information to feature.xml. Probably in order to not have conflicts in feature.xml I would need to pull + rebase and then add my binding information to the feature.xml. My guess is that someone added something new there and now it conflicts with my information. Probably would be easy to merge also...

martinvw commented:
I think the best solution is to recreate your initial commit

```
git reset HEAD~22
```

This does a soft reset so all the changes of your commits should still be there.

Then you should revert all the changes to the `feature.xml` after that you create a new commit for your binding and force push that. After that you should rebase you branch on the current openhab/master and add your changes to feature.xml. 

Please check whether it does not accidentally change the whole file again... afaik this should normally not happen git is able to handle linux vs windows line ends correct normally.

theater commented:
OK. Thanks for pointing that out. I finally managed to fix it by reverting back the file to latest state and editting it with notepad++ using unix line endings. Now it looks OK. At least it does not show any merge warnings. 
Now another issue appeared with this DCO bot - it says that I did not signoff my last two commits but I see the signoff text in them. Is it a bug in DCO or again some formatting issue?

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/EvoCommunicator.java:
```diff
@@ -0,0 +1,324 @@
+     * An array of an array of the raw bytes received from the TCP/IP
+     * stream.
+     *
+     * @throws ParadoxBindingException
```
This javadoc is incomplete. It misses the param and other throws.

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/GenericCommunicator.java:
```diff
@@ -0,0 +1,331 @@
+
+    protected final Logger logger = LoggerFactory.getLogger(GenericCommunicator.class);
+
+    protected Socket socket;
```
Do these fields need to protected. And some of them could be declared `final` I guess?

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/GenericCommunicator.java:
```diff
@@ -0,0 +1,331 @@
+                    "Unable to sleep thread during socket close phase. Could lead to issues if reconnect occurs. {}",
+                    e);
+        } catch (IOException e) {
+            logger.warn("IO exception during socket/stream close operation. {}", e);
```
Only put exception as last param, but don't refer to it in the message string. This is the way to do it for exception logging.
```suggestion
            logger.warn("IO exception during socket/stream close operation. ", e);
```

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/Main.java:
```diff
@@ -0,0 +1,145 @@
+            cache.setPartitionLabels(communicator.readPartitionLabels());
+            cache.setZoneLabels(communicator.readZoneLabels());
+        }
+        // logger.error("Communicator cannot refresh cached memory map. Exception: ", e);
```
Remove commented out code

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/messages/HeaderCommand.java:
```diff
@@ -0,0 +1,48 @@
+
+/**
+ * {@link HeaderCommand}
+ * From Jean's excel sheet:
```
Who is Jean? No problem it's stated here. But others don't know what this refers to so it might help to put in some context. This might help future maintainers.

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/EntityBaseHandler.java:
```diff
@@ -0,0 +1,86 @@
+ */
+public abstract class EntityBaseHandler extends BaseThingHandler {
+
+    private static final long INITIAL_DELAY = 10;
```
Put time unit in constant name so it's clear what the value represents.
```suggestion
    private static final long INITIAL_DELAY_SECONDS = 10;
```

Hilbrand commented on features/karaf/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -582,6 +582,11 @@
         <bundle start-level="80">mvn:org.openhab.binding/org.openhab.binding.orvibo/${project.version}</bundle>
     </feature>
 
+	<feature name="openhab-binding-paradox" description="Paradox Alarm Binding" version="${project.version}">
```
Fix indentation. Should be spaces.

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/model/Zone.java:
```diff
@@ -0,0 +1,44 @@
+        logger.trace("Zone {} state updated to:\tOpened: {}, Tampered: {}, LowBattery: {}",
+                new Object[] { getLabel(), zoneState.isOpened(), zoneState.isTampered(), zoneState.hasLowBattery() });
+    }
+}
```
Missing new line at end of file

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/EvoCommunicator.java:
```diff
@@ -0,0 +1,324 @@
+            for (int i = 1; i <= 60; i++) {
+                result.add(readZoneLabel(i));
+            }
+        } catch (Exception e) {
```
Only catch specific exception (or `RuntimeException` if you expect that.

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxAlarmHandlerFactory.java:
```diff
@@ -0,0 +1,77 @@
+            try {
+                logger.debug("createHandler(): ThingHandler created for {}", thingTypeUID);
+                return new ParadoxIP150BridgeHandler((Bridge) thing);
+            } catch (Exception e) {
```
Only catch specific exception (or `RuntimeException` if you expect that.

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java:
```diff
@@ -0,0 +1,281 @@
+            updateStatus(ThingStatus.OFFLINE);
+
+            scheduler.execute(this::initializeCommunication);
+        } catch (Exception e) {
```
Only catch specific exception (or `RuntimeException` if you expect that.

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java:
```diff
@@ -0,0 +1,281 @@
+                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE,
+                                "Bringing bridge offline due to reinitialization of communicator.");
+                        resetCommunicator();
+                    } catch (Exception e) {
```
Only catch specific exception (or `RuntimeException` if you expect that.

Hilbrand commented:
I have no idea why the DCO fails. It looks like the sign off is ok.

To solve the conflicts. Do a `git pull --rebase`. Fix the conflict and then a `git push --force-with-lease` to put the changes back.

Hilbrand commented on addons/binding/org.openhab.binding.paradoxalarm/ESH-INF/thing/ip150connector.xml:
```diff
@@ -0,0 +1,54 @@
+	xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+
+	<!-- Sample Thing Type -->
```
Left over from sample xml

Hilbrand commented on bom/openhab-addons/pom.xml:
```diff
@@ -535,6 +535,11 @@
       <artifactId>org.openhab.binding.orvibo</artifactId>
       <version>${project.version}</version>
+    </dependency>
+	<dependency>
```
indentation

theater commented:
OK I think I've merged everything and migrated to the new build system. I hope I didn't mess up any metadata. Please review and let me know if anything else is required.

Hilbrand commented:
Migration looks fine. It all looks ok. Only one thing left. You need to add your binding to the `CODEOWNERS` file.

theater commented:
> Migration looks fine. It all looks ok. Only one thing left. You need to add your binding to the `CODEOWNERS` file.

Done. Thought that I did it before... maybe I've lost the change during last changes related to flattening.

Hilbrand commented:
@martinvw Can you have a second look at this new paradoxalarm binding as you did an earlier review.

Hilbrand commented:
Jenkins build seems to fail on another binding, but this new binding succeeds in Jenkins.

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/GenericCommunicator.java:
```diff
@@ -0,0 +1,332 @@
+            initializeSocket();
+        }
+
+        logger.debug("Step1");
```
If you can split up the code in steps, please split it up in methods.

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/GenericCommunicator.java:
```diff
@@ -0,0 +1,332 @@
+        sendPacket(ipPacket);
+        byte[] loginPacketResponse = receivePacket();
+        if (!isInialLoginSuccessful(loginPacketResponse)) {
+            // logoutSequence();
```
Still needed?

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/GenericCommunicator.java:
```diff
@@ -0,0 +1,332 @@
+            logoutSequence();
+        }
+        Thread.sleep(300);
+        // TODO check why after a short sleep, a 37 bytes packet is received after logon
```
There is a TODO, what should be done and can it be completed before merging the PR?

theater replied:
I don't care about this 37 bytes packet anymore. It works without knowing what exactly it does. :)

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/GenericCommunicator.java:
```diff
@@ -0,0 +1,332 @@
+                if (payloadResponseByte == 0x00) {
+                    logger.info("Login - Login to IP150 - OK");
+                    return true;
+                }
```
Missing break!

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/GenericCommunicator.java:
```diff
@@ -0,0 +1,332 @@
+        }
+
+        switch (payloadResponseByte) {
+            case 0x01:
```
Missing breaks!!!

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/GenericCommunicator.java:
```diff
@@ -0,0 +1,332 @@
+                    return Arrays.copyOfRange(result, 0, result[1] + 16);
+                }
+            } catch (IOException e) {
+                logger.debug("Unable to retrieve data from RX. {}", e.getMessage());
```
Are you never interested in the stacktrace? Would it make sense to not use a placeholder and just pass e

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/GenericCommunicator.java:
```diff
@@ -0,0 +1,332 @@
+                }
+            } catch (IOException e) {
+                logger.debug("Unable to retrieve data from RX. {}", e.getMessage());
+                Thread.sleep(200);
```
Please use scheduler

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/GenericCommunicator.java:
```diff
@@ -0,0 +1,332 @@
+                }
+            }
+        }
+        throw new IOException("Unable to read from socket or received data is wrong.");
```
Shouldn’t you write the code a little bit different such that you throw the original exception instead for example by inspecting the retry counter in the catch.

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/GenericCommunicator.java:
```diff
@@ -0,0 +1,332 @@
+        return isOnline;
+    }
+
+}
```
Missing new line at the end of the file

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/Main.java:
```diff
@@ -0,0 +1,142 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link Main} - used for testing purposes of low-level stuff.
```
Should it be in a testing package then?

martinvw replied:
Does it belong in this PR? This does not follow the guidelines of the code so puts out a very wrong example for other binding developers. We should not catch generic exceptions, provide context when logging, never use system exit. Please let me know what the target audience is for this class

theater replied:
The target for this class is to be used for testing the low-level live communication. That's why all the ugly mechanics. I've had the same proposal by Hilbrand but I would strongly prefer to keep this class just for testing outside of openhab environment. Of course probably I can try to keep it in separate branch but that would be overhead. Unless it's strictly forbidden to use main class for testing I would like to keep it. Please let me know if we can agree on keeping it.

Hilbrand replied:
We the new build system it's now allowed to have a test folder `/src/test/java` in the same project. Maybe you can move this file to that location so it's outside the main code.

theater replied:
Now when looking into the options for fixing Thread.sleep() requirement by Martin, I believe I will have to get rid of main method simply because I will need to make dependency to ThreadPoolManager in my own model... I don't like this at all because I wanted to keep the low-level part simple and loosly coupled from the OpenHAB framework. Maybe I can check if scheduler is null to create my own implementation but I'm not sure if it worths the effort just for the sake of loose coupling of the model. What do you think @Hilbrand  @martinvw ?

martinvw replied:
Maybe I lost a little bit of recent updates but afaik all you need is the scheduler of the parent class of the handler. If really need you could get your instance of ScheduledExecutorService from the jdk for your main class.

@Hilbrand please correct me if I’m wrong / outdated.

theater replied:
Yeah it's done in BaseThingHandler. There is initialized the scheduler using static call to this: ThreadPoolManager
            .getScheduledPool(THING_HANDLER_THREADPOOL_NAME). Probably somewhere I need to provide the executor service to my communicator and in main class file I'll see how to get it from jdk. Will look into that further tomorrow. Thanks a lot for the feedback ! Most of feedback is valuable for me. It's good that I'm home these days so I can have a look into the feedback and do the fixes faster than normal...

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/messages/EpromRequestPayload.java:
```diff
@@ -0,0 +1,48 @@
+    @Override
+    protected byte calculateControlByte() {
+        int address = getAddress();
+        logger.trace("Address: {}", String.format("0x%02X,\t", address));
```
Consider wrapping this in a check whether trace is enabled the string format is not cheap / free

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/messages/MemoryRequestPayload.java:
```diff
@@ -0,0 +1,74 @@
+        this.address = address;
+        this.bytesToRead = bytesToRead;
+
+        logger.trace("MessageStart: {}", String.format("0x%02X,\t", MESSAGE_START));
```
See above

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/messages/ParadoxIPPacket.java:
```diff
@@ -0,0 +1,191 @@
+    }
+
+    public ParadoxIPPacket(String payload, boolean isChecksumRequired) throws IOException {
+        this(payload.getBytes("US-ASCII"), isChecksumRequired);
```
See earlier remark about standardcharsets and check all,

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/messages/ParadoxIPPacket.java:
```diff
@@ -0,0 +1,191 @@
+            this.payloadLength = (short) payload.length;
+        }
+
+        // TODO: Figure out how to fill up to 16, 32, 48, etc sizes with 0xEE
```
There is a TODO, what should be done and can it be completed before merging the PR?

theater replied:
This is not mandatory feature but it's good to implement it as per reverse engineering document by the guy who initially did the RE.

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/exceptions/ParadoxBindingException.java:
```diff
@@ -0,0 +1,45 @@
+    private static final long serialVersionUID = -5771699322577106346L;
+
+    public ParadoxBindingException() {
+        super();
```
Not needed

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/exceptions/ParadoxBindingException.java:
```diff
@@ -0,0 +1,45 @@
+package org.openhab.binding.paradoxalarm.internal.exceptions;
+
+/**
+ * The {@link ParadoxBindingException} Wrapper of Exception class. May not be needed.
```
May not be needed? Is it needed or not in hind sight?

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/EntityBaseHandler.java:
```diff
@@ -0,0 +1,86 @@
+            if (ThingStatus.ONLINE == getThing().getStatus()) {
+                updateEntity();
+            } else {
+                logger.debug("Received REFRESH command but {} is OFFLINE", getThing().getUID());
```
Is there a reason why you should not try anyway? Maybe it’s online now?

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/EntityBaseHandler.java:
```diff
@@ -0,0 +1,86 @@
+        logger.trace("Start initializeDelayed() in {}", getThing().getUID());
+        try {
+            ParadoxPanel panel = ParadoxPanel.getInstance();
+            if (!panel.isPanelSupported()) {
```
Where do host not founds and other communication errors end up?

theater replied:
Everything ends up in the catch clause of initializeCommunication() method. If something goes wrong the bridge status gets updated to offline with the error from the catched exception.

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java:
```diff
@@ -0,0 +1,276 @@
+            panelTypeStr = config.getPanelType();
+        }
+
+        logger.info("Phase2 - Creating communicator for panel {}", panelType);
```
Please leave this out by default bindings should be silent.

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java:
```diff
@@ -0,0 +1,276 @@
+    public void initialize() {
+        logger.debug("Start initialize()...");
+        updateStatus(ThingStatus.UNKNOWN);
+        logger.info("Starting creation of communicator.");
```
Please check all

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java:
```diff
@@ -0,0 +1,276 @@
+                updateStatus(ThingStatus.ONLINE);
+            }
+        } catch (IOException | InterruptedException | ParadoxBindingException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
```
See above communication or configuration error

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java:
```diff
@@ -0,0 +1,276 @@
+
+            if (failedReadAttempts > 0) {
+                logger.info("Successfully refreshed memory map after {} failed attempts.", failedReadAttempts);
+                failedReadAttempts = 0;
```
Should we update the status here?

theater replied:
The status handling is done via the handleSocketReadError method. I would prefer to keep trying without updating thing status until the treshold is reached, otherwise all things start flapping down/up which makes a lot of changes. Not sure if it's good idea. Anyway the things are getting all the data from the last cached result. If something goes really wrong they will get offline after reaching there bridge retry treshold together with the bridge. If it's something temporary I keep the status online. Is that wrong that we have certain amount of time of false behavior saying it's online while it's actually retrying?

martinvw replied:
That sounds good, thanks for the explanation

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java:
```diff
@@ -0,0 +1,276 @@
+
+        } catch (IOException e) {
+            handleSocketReadError(e);
+        } catch (InterruptedException | ParadoxBindingException e) {
```
If we receive this InterruptedException, the interrupted flag was cleared and we most likely need to reset the flag:

http://www.yegor256.com/2015/10/20/interrupted-exception.html
https://www.ibm.com/developerworks/library/j-jtp05236/

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java:
```diff
@@ -0,0 +1,276 @@
+            }
+
+        } catch (IOException e) {
+            handleSocketReadError(e);
```
Should we mark it offline?

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java:
```diff
@@ -0,0 +1,276 @@
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.debug("Received command {}", command.toFullString());
+        if (ThingStatus.OFFLINE.equals(getThing().getStatus()) && command instanceof RefreshType) {
```
The thingstatus is an enumeration you can use ==

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java:
```diff
@@ -0,0 +1,276 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.debug("Received command {}", command.toFullString());
```
Do you really need the full string now you will always (calculate) the full string even if the logging does not apply

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxPanelHandler.java:
```diff
@@ -0,0 +1,73 @@
+                        panelInformation.getBootLoaderVersion().toString());
+            }
+        } catch (ParadoxBindingException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "Unable to retrieve ParadoxPanel instance. " + e);
```
Rather only add the e.getMessage

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxPartitionHandler.java:
```diff
@@ -0,0 +1,58 @@
+                        new StringType(partition.getState().getAdditionalState()));
+            }
+        } catch (ParadoxBindingException e) {
+            logger.warn("Unable to update partition with Id {} due to missing ParadoxPanel. Exception: {}",
```
Does index + 1 also work or do you really need the string conversion

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/model/Entity.java:
```diff
@@ -0,0 +1,57 @@
+
+    @Override
+    public String toString() {
+        return "Entity [id=" + id + ", label=" + label;
```
It looks inconsistent seems to be missing a closing ] 😊

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/model/PanelType.java:
```diff
@@ -0,0 +1,63 @@
+        byte[] panelTypeBytes = Arrays.copyOfRange(infoPacket, 6, 8);
+        String key = "0x" + HexUtils.bytesToHex(panelTypeBytes);
+
+        return ParadoxInformationConstants.panelTypes.getOrDefault(key, PanelType.UNKNOWN);
```
That is a strange call between multiple classes, why does this method not live next to the map or vice versa?

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/model/ParadoxInformationConstants.java:
```diff
@@ -0,0 +1,46 @@
+ * @author Konstantin_Polihronov - Initial contribution
+ */
+public class ParadoxInformationConstants {
+    public static Map<String, PanelType> panelTypes = new HashMap<String, PanelType>();
```
Please use the diamond operator, ie replace the second occurrence of the same generic by `<>`

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/model/ParadoxPanel.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    private List<Zone> createZones() {
+        zones = new ArrayList<Zone>();
```
Diamond operator pleAse check all

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/model/Partition.java:
```diff
@@ -0,0 +1,44 @@
+ */
+public class Partition extends Entity {
+
+    private static Logger logger = LoggerFactory.getLogger(Partition.class);
```
Can you make it `final` instead of `static`?

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/model/PartitionState.java:
```diff
@@ -0,0 +1,323 @@
+
+    public String getAdditionalState() {
+        StringBuilder sb = new StringBuilder();
+        // TODO check if isInAlarm also includes the other three if yes -> check only the other three
```
There is a TODO, what should be done and can it be completed before merging the PR?

theater replied:
Removed. It's definitely not something important right now...

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/model/Zone.java:
```diff
@@ -0,0 +1,44 @@
+
+    public void setZoneState(ZoneState zoneState) {
+        this.zoneState = zoneState;
+        logger.trace("Zone {} state updated to:\tOpened: {}, Tampered: {}, LowBattery: {}",
```
Why the new object array?

theater replied:
leftover from my previous implementation with different logger I guess. :)

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/util/ParadoxUtil.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    public static void printPacket(String description, byte[] array) {
+        logger.trace("Packet payload size: {}", array[1]);
```
Should this method be completely wrapped in a check whether trace logging is enabled?

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/resources/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+
+	<name>ParadoxAlarm Binding</name>
+	<description>This is the binding for ParadoxAlarm.</description>
+	<author>Konstantin_Polihronov</author>
```
Can you remove the underscore from your name everywhere?

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/resources/ESH-INF/thing/ip150connector.xml:
```diff
@@ -0,0 +1,54 @@
+	xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+
+	<!-- Sample Thing Type -->
```
Sample?

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/resources/ESH-INF/thing/ip150connector.xml:
```diff
@@ -0,0 +1,54 @@
+			</parameter>
+			<parameter name="ipAddress" type="text" required="true">
+				<label>IP Address</label>
+				<description>IP address of IP150</description>
```
Can this only be an ip or just a hostname?

theater replied:
I never looked into the low level implementation of java sockets but I would expect that if it's resolvable hostname, it should also work...

martinvw replied:
Me too

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/resources/ESH-INF/thing/ip150connector.xml:
```diff
@@ -0,0 +1,54 @@
+			</parameter>
+			<parameter name="ipAddress" type="text" required="true">
+				<label>IP Address</label>
+				<description>IP address of IP150</description>
```
Please add:

```
<context>network-address</context>
```

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/resources/ESH-INF/thing/panel.xml:
```diff
@@ -0,0 +1,70 @@
+	xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+
+	<!-- Sample Thing Type -->
```
Sample

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/resources/ESH-INF/thing/panel.xml:
```diff
@@ -0,0 +1,70 @@
+
+ 		<channels>
+			<channel id="state" typeId="panelState" />
+			<!-- <channel id="serialNumber" typeId="serialNumber" />
```
Can you remove those

theater replied:
Hm... those were removed upon request by Hilbrand... it seems I've reverted some of the last changes. I need to get back to Hilbrands latest requests and verify everything is fixed again. :(

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/resources/ESH-INF/thing/panel.xml:
```diff
@@ -0,0 +1,70 @@
+		<description>Shows the state of the communication to panel (online/offline)</description>
+		<state readOnly="true" pattern="%s" />
+	</channel-type>
+	<channel-type id="serialNumber">
```
Can these channels now be removed?

theater commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/EvoCommunicator.java:
```diff
@@ -0,0 +1,327 @@
+        List<String> result = new ArrayList<>();
+
+        try {
+            for (int i = 1; i <= 60; i++) {
```
I planned to extract all constants relevant to particular Paradox system in a separate PR as a feature addon/fix. Maybe inside the ENUM which holds the types, or maybe in a separate map from with system type as a key and static Paradox feature struct object as value. If you insist on this I can try to implement it now but prefer not to touch too many things right now as it's very stable.

martinvw replied:
That is fine

theater commented:
Hm... I've been looking at the coding few minutes ago. I understand the issues with Thread.sleep() and completely agree that it needs to be changed ...but If we want to get rid of all Thread.sleep() calls I will have to completely redevelop the communication part to be asynchronous using callables and futures which I never used before. I'm not sure how much time it will take and what is the overall risk of changing this for the binding itself. Overall I agree though that with current approach the threads used in my binding can become laggy and can lock up the whole application for some time if there are other bindings using simmilar sleeps. I'm really not sure how to continue with that right now... From PR perspective how should I approach that in order not to flood with changes the PR? New branch?

theater commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/EvoCommunicator.java:
```diff
@@ -0,0 +1,327 @@
+                    logger.warn("Command {} not implemented.", command);
+            }
+        } catch (IOException | InterruptedException e) {
+            logger.warn("Error while executing command {}. Exception:{}", command, e);
```
I think this has been answered with the other explanation about the IO exceptions entry point. They lead to offline status of the bridge but it is not handled in the backend model. OK?

martinvw replied:
I think my point is still valid, because you do not distinguish between catching just any IOException or the UnknownHostException.

theater commented:
Hi,
when you have time could you please review the receivePacket() and asyncReceivePacketHandle() methods in generic communicator? 
I got rid of the Thread.sleep() but I'm not sure if calling future.get() is not doing the same because I've read in OCP exam guide that when you call get() the current thread waits for the other thread execution to finish and provide the future<result>. Not sure if this thread is reusable while waiting and if there is some callback mechanism implemented so I'm a bit unsure if I did it right.

Cheers,
K.

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/GenericCommunicator.java:
```diff
@@ -197,10 +205,13 @@ private void step7FinalInitializationRequest(byte[] initializationMessage)
-        receivePacket();
+
+        try {
+            scheduler.schedule(this::receivePacket, 300, TimeUnit.MILLISECONDS).get();
```
Because of the `get  it does exactly the same, but why do you / would you need the get?

theater replied:
I'd like not to move forward unless I have received this "ghost" 37 byte packet which I'm not aware what exactly does. :) Otherwise this packet will get received together with next call and will mess up the normal communication. Since this is relevant only for the login sequence and usually happens only during initialization/reset I believe it's not a big deal even with Thread.sleep().

martinvw replied:
Sorry, but I think that is really error prone and not good. Because this method is indirectly called from the constructor so you are blocking a thread, still effectively almost the same as the sleep which we had imho, inside a constructor. Cant you just implement some logic to ignore this ghost packet whenever it arrives? Or just ignore the first packet after a login. 

@openhab/2-x-add-ons-maintainers do you have any great ideas to resolve this (quickly)

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/test/java/main/Main.java:
```diff
@@ -68,8 +75,7 @@ public static void main(String[] args) {
-            Thread.sleep(5000);
-            paradoxSystem.updateEntitiesStates();
+            scheduler.schedule(paradoxSystem::updateEntitiesStates, 5000, TimeUnit.MILLISECONDS);
         } catch (InterruptedException e1) {
```
Is this catch still valid / needed?

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/test/java/main/Main.java:
```diff
@@ -68,8 +75,7 @@ public static void main(String[] args) {
             updateDataCache(communicator, false);
-            Thread.sleep(5000);
-            paradoxSystem.updateEntitiesStates();
+            scheduler.schedule(paradoxSystem::updateEntitiesStates, 5000, TimeUnit.MILLISECONDS);
```
5 seconds?

theater replied:
That's basically the each 5 seconds call to the system to retrieve new states. 5 seconds is OK, thread sleep is also OK here as this is the main class for testing and no other threads will do anything concurent but moved to scheduler just to test if it works also here

martinvw replied:
Sorry that i was a little bit vague in this comment I meant that it make more sense to code it as 5 seconds instead of 5000 msec

martinvw commented on bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/GenericCommunicator.java:
```diff
@@ -257,28 +268,40 @@ private void sendPacket(byte[] packet) throws IOException {
     }
 
-    // TODO think about better approach
     protected byte[] receivePacket() throws InterruptedException, IOException {
```
Not completely, but close

I would expect something a little bit different

maybe a field `consecutiveErrors`

```java
protected byte[] receivePacket() throws IOException {
try {
byte[] result = new byte[256];
            rx.read(result);
            ParadoxUtil.printPacket("RX:", result);
            if (result[1] > 0 && result[1] + 16 < 256) {
                return processResponse(Arrays.copyOfRange(result, 0, result[1] + 16));
            }
} catch (...){
}
if (consecturiveErrors < 4) {
 // reschedule receivePacket
resultFuture = scheduler.schedule(receivePacketCallable, 200, TimeUnit.MILLISECONDS);
}

```

Also consider closing / cancelling Future's those from the close method.

martinvw commented:
@theater I made a very rough sketch of my suggestion if its not clear then just ping.

Then the main thing is that if you want to have it retryable it should be responsible of processing the results itself either by directly calling some method or maybe by receiving some callable. It will not work if you depend on a return value, then the caller has to wait in-sync.

theater commented:
Hi Martin, thanks for advice. Do you mean to implement something like a listener for the receive part? If so I will have to think about how to approach that. Especially I'm not quite sure how to do it in login case where we have a sequence of particular requests/responses...

martinvw commented:
> Hi Martin, thanks for advice. Do you mean to implement something like a listener for the receive part? If so I will have to think about how to approach that. Especially I'm not quite sure how to do it in login case where we have a sequence of particular requests/responses...

Yes that is indeed the common way to do such communication. if the request / responses are specific for the login sequence you could keep track of the state of the sign-in. It would become a little like a state machine then.

theater commented:
Thanks a lot for the help in making the code better !

Just for information to anyone who may be concerned - I will not have time in the next 2-3 or maybe 4 weeks to do the redesign and refactoring of this pull request because of other time consuming tasks outside of OpenHAB community efforts. Hopefully in the background I'll get with a nice solution in my mind for the asynchronous communication design. :)

When I have time to continue I will push changes and will notify reviewers.

Cheers,
K.

theater commented:
Hi,
I've had pretty busy two months so I apologize for not being able to finalize the code changes.
Due to the initial design I had to change a lot of stuff which took me more time than expected. I would propose that a more thorough review will be required.

The more major changes so far are:
- I've created an enum called CommunicationState which acts as a state machine mostly for the logon process.
- Implemented a request/response classes which make it possible to send back to the callers the response based on the request specifics, i.e. particular RAM page, EPROM zone or a logon packet.
- Implemented a synchronization queue. All requests go into sendQueue and upon send are popped from send queue and are pushed into receiveQueue. Due to nature of Paradox communication receive queue is with priority, i.e. if there is anything in receive queue we attempt to read the socket first and only after receive queue is empty then we attempt to send. We never send any packet if we have something to read.. 
- Changed some exceptions handling by wrapping them in runtime exceptions.

Cheers,
K.

martinvw commented:
> P.S. Just figured that I need to add some bridge status changes in the bridge handler in the cases when communication broken in some way. Will do that in next commit.

Is that still pending or is that the commit you did after that message

martinvw commented on features/karaf/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -1,1192 +0,0 @@
-﻿<?xml version="1.0" encoding="UTF-8"?>
```
This is strange to have as a diff, did you rebase properly?

theater replied:
Hm.... I think what I did is I checked if the file exists in master branch of main repository and when I didn't find it, I deleted it. Also I believe the latest push is not rebased to current master. Maybe I should pull master and rebase again...

martinvw replied:
Nevermind git was able to sort it out :-)

Thanks for your hard work!

theater commented:
> > P.S. Just figured that I need to add some bridge status changes in the bridge handler in the cases when communication broken in some way. Will do that in next commit.
> 
> Is that still pending or is that the commit you did after that message
It is pushed with the commit after my comment.
